crypto/crypto.go

```diff
@@ -6,112 +6,120 @@ import (
 	"github.com/relab/hotstuff/consensus"
 )
 
-type base struct {
-	consensus.CryptoImpl
+type crypto struct {
+	mods *consensus.Modules
+	consensus.CryptoBase
 }
 
-// New returns a new base implementation of the Crypto interface. It will use the given CryptoImpl to create and verify
+// New returns a new implementation of the Crypto interface. It will use the given CryptoBase to create and verify
 // signatures.
-func New(impl consensus.CryptoImpl) consensus.Crypto {
-	return base{CryptoImpl: impl}
+func New(impl consensus.CryptoBase) consensus.Crypto {
+	return &crypto{CryptoBase: impl}
 }
 
 // InitConsensusModule gives the module a reference to the Modules object.
 // It also allows the module to set module options using the OptionsBuilder.
-func (base base) InitConsensusModule(mods *consensus.Modules, cfg *consensus.OptionsBuilder) {
-	if mod, ok := base.CryptoImpl.(consensus.Module); ok {
+func (c *crypto) InitConsensusModule(mods *consensus.Modules, cfg *consensus.OptionsBuilder) {
+	c.mods = mods
+	if mod, ok := c.CryptoBase.(consensus.Module); ok {
 		mod.InitConsensusModule(mods, cfg)
 	}
 }
 
 // CreatePartialCert signs a single block and returns the partial certificate.
-func (base base) CreatePartialCert(block *consensus.Block) (cert consensus.PartialCert, err error) {
-	sig, err := base.Sign(block.Hash())
+func (c crypto) CreatePartialCert(block *consensus.Block) (cert consensus.PartialCert, err error) {
+	sig, err := c.Sign(block.ToBytes())
 	if err != nil {
 		return consensus.PartialCert{}, err
 	}
 	return consensus.NewPartialCert(sig, block.Hash()), nil
 }
 
 // CreateQuorumCert creates a quorum certificate from a list of partial certificates.
-func (base base) CreateQuorumCert(block *consensus.Block, signatures []consensus.PartialCert) (cert consensus.QuorumCert, err error) {
+func (c crypto) CreateQuorumCert(block *consensus.Block, signatures []consensus.PartialCert) (cert consensus.QuorumCert, err error) {
 	// genesis QC is always valid.
 	if block.Hash() == consensus.GetGenesis().Hash() {
 		return consensus.NewQuorumCert(nil, 0, consensus.GetGenesis().Hash()), nil
 	}
-	sigs := make([]consensus.Signature, 0, len(signatures))
+	sigs := make([]consensus.QuorumSignature, 0, len(signatures))
 	for _, sig := range signatures {
 		sigs = append(sigs, sig.Signature())
 	}
-	sig, err := base.CreateThresholdSignature(sigs, block.Hash())
+	sig, err := c.Combine(sigs...)
 	if err != nil {
 		return consensus.QuorumCert{}, err
```

Comment:
We should avoid returning empty structs when we have error; it causes unnecessary allocation.

Reply:
I thought that the memory would be allocated on the stack since it is returned by value?

Reply:
I have gotten into the habit of always returning a pointer type in such scenarios. I'm not sure if it always doesn't escape to the heap, but it probably doesn't matter much in terms of performance. I had to look it up, and it seems both are okay, but it seems the [recommended pattern](https://morioh.com/p/385f6171e967) is to use the `nil` variant. This seems to be based on the advice on [Declaring Empty Slices](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices), which also references a talk about nil that I've watched before but didn't rewatch.

I guess one argument for using a `nil` pointer is that it becomes easier to identify/see that it is in fact returning a `nil` as opposed to examining the code and seeing `consensus.QuorumCert{}` you sort of have to think what it means that this is non-nil.

Reply:
I just don't think that it is a good idea for constructor functions to return a pointer unless it is needed for some specific purpose. What good could a pointer do in this case?